docs/glossary.md

```diff
@@ -0,0 +1,204 @@
+
+Similar to a branch conflict, when a [change](#change) is rewritten locally
+and remotely, for example, then the change will be in a conflicted state. We
+call that a [divergent change](#divergent-change).
+
+## Divergent change
+
+A divergent change is a [change](#change) that has more than one
+[visible commit](#visible-commits).
+
+## Head
+
+A head is a commit with no descendants. The [view](#view) records which
+anonymous heads (heads without a branch pointing to them) are visible at a
+given [operation](#operation). Note that this is quite different from Git's
+[HEAD](https://git-scm.com/book/en/v2/Git-Internals-Git-References#ref_the_ref).
+
+## Operation
+
+A snapshot of the [visible commits](#visible-commits) and [branches](#branches)
+at a given point in time (technically a [view object](#view)), together with
+some metadata. The metadata includes the username, hostname, timestamps, and
+pointers to the operation's parents.
+
+## Operation log
+
+The operation log is the
+[DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) formed by
+[operation](#operation) objects, much in the same way that commits form a DAG,
+which is sometimes called the "commit history". 
+
+## Repository
+
+Basically everything under `.jj/`, i.e. the full set of [operations](#operation)
+and [commits](#commit).
+
+## Remote
+
+TODO
+
+## Revision
+
+A synonym for [Commit](#commit).
+
+## Revset
+
+Jujutsu supports a functional language for selecting a set of revisions.
+Expressions in this language are called "revsets". See [here](revsets.md) for
+details. We also often use the term "revset" for the set of revisions selected
+by a revset.
+
+## Rewrite
+
+To "rewrite" a commit means to create a new version of that commit with
+different contents, metadata (including parent pointers), or both. Rewriting a
+commit results in a new commit, and thus a new [commit ID](#commit-id), but the
+[change ID](#change-id) generally remains the same.
+
+## Root commit
+
+The root commit is a virtual commit at the root of every repository. It has a
+commit ID consisting of all '0's (`00000000...`) and a change ID consisting of
+all 'z's (`zzzzzzzz...`). It can be referred to in [revsets](#revset) by the
+special name `root`. Note that our definition of "root commit" is different from
+Git's; Git's "root commits" are the first commit(s) in the repository, i.e. the
+commits `jj log -r root+` will show.
+
+## Tree
+
+A tree object represents a snapshot of a directory in the repository. Tree
+objects are defined recursively; each tree object only has the files and
+directories contained directly in the directory it represents.
+
+## Visible commits
+
+Visible commits are the commits you see in `jj log -r 'all()'`. They are the
+commits that are reachable from an anonymous head in the [view](#view).
+Ancestors of a visible commit are implicitly visible.
+
+## View
+
+A view is a snapshot of branches and their targets, anonymous heads,
```

Comment:
I would put (visible), with or without parantheses, before anonymous heads here. This is related to a previous comment.

Reply:
It's actually only the "anonymous heads" that define visibility. They're actually called just "heads" in the model (https://github.com/martinvonz/jj/blob/cf402af9c15755064cc3cc435b2c02223e51354d/lib/src/op_store.rs#L193). "Visible heads" is technically more correct than "anonymous heads", but I found it hard to explain it without ending up with circular definitions. See some earlier discussion between me and @chooglen in this PR. Perhaps we should change back to "visible heads" if we can think of a better way to explain it.

Reply:
Sounds good to postpone fixing this, once we figure out a better way. I think what you have now is good enough for people who care about such details to figure it out after a bit of confusion, and most people will only care about the gist of it anyway.